File: pr048.hs
{-
The series, 1**1 + 2**2 + 3**3 + ... + 10**10 = 10405071317.

Find the last ten digits of the series, 1**1 + 2**2 + 3**3 + ... + 1000**1000.
-}

module Main where

numLast :: Int -> Integer -> Integer
numLast n x = read $ drop dropLen numStr
              where
                numStr = show x
                dropLen = length numStr - n

numLast10 = numLast 10

selfPowered :: Integral a => a -> a
selfPowered x = x ^ x

main = print $ numLast10 $ sum $ map selfPowered [1..1000]

File: pr048.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
The series, 1**1 + 2**2 + 3**3 + ... + 10**10 = 10405071317.

Find the last ten digits of the series, 1**1 + 2**2 + 3**3 + ... + 1000**1000.
"""

def main():
  """main function"""
  sum = reduce(lambda a, b: a + b,
               map(lambda x: x ** x, range(1, 1001)))
  print int(str(sum)[-10:])


if __name__ == "__main__":
  main()
